src/interfaces/chain.h

```diff
@@ -0,0 +1,251 @@
+#ifndef BITCOIN_INTERFACES_CHAIN_H
+#define BITCOIN_INTERFACES_CHAIN_H
+
+#include <optional.h>               // For Optional and nullopt
+#include <policy/rbf.h>             // For RBFTransactionState
+#include <primitives/transaction.h> // For CTransactionRef
+
+#include <memory>
+#include <stddef.h>
+#include <stdint.h>
+#include <string>
+#include <vector>
+
+class CBlock;
+class CRPCCommand;
+class CReserveScript;
+class CScheduler;
+class CValidationState;
+class UniValue;
+class uint256;
+struct CBlockLocator;
+struct FeeCalculation;
+
+namespace interfaces {
+
+class Handler;
+
+//! Interface for giving wallet processes access to blockchain state.
+class Chain
+{
+public:
+    virtual ~Chain() {}
+
+    //! Interface for querying locked chain state, used by legacy code that
+    //! assumes state won't change between calls. New code should avoid using
+    //! the Lock interface and instead call higher-level Chain methods
+    //! that return more information so the chain doesn't need to stay locked
+    //! between calls.
+    class Lock
+    {
+    public:
+        virtual ~Lock() {}
+
+        //! Get current chain height, not including genesis block (returns 0 if
+        //! chain only contains genesis block, nothing if chain does not contain
+        //! any blocks).
+        virtual Optional<int> getHeight() = 0;
+
+        //! Get block height above genesis block. Returns 0 for genesis block,
+        //! for following block, and so on. Returns nothing for a block not
+        //! included in the current chain.
+        virtual Optional<int> getBlockHeight(const uint256& hash) = 0;
+
+        //! Get block depth. Returns 1 for chain tip, 2 for preceding block, and
+        //! so on. Returns 0 for a block not included in the current chain.
+        virtual int getBlockDepth(const uint256& hash) = 0;
+
+        //! Get block hash.
+        virtual uint256 getBlockHash(int height) = 0;
+
+        //! Get block time.
+        virtual int64_t getBlockTime(int height) = 0;
+
+        //! Get block median time past.
+        virtual int64_t getBlockMedianTimePast(int height) = 0;
+
+        //! Check if block is empty.
+        virtual bool blockHasTransactions(int height) = 0;
+
+        //! Return height of earliest block in chain with timestamp equal or
+        //! greater than the given time, or nothing if there is no block with a
+        //! high enough timestamp.
+        virtual Optional<int> findEarliestAtLeast(int64_t time) = 0;
+
+        //! Return height of last block in chain with timestamp less than the
+        //! given, and height less than or equal to the given, or nothing if
```

Comment:
Shouldn't this be "height greater than or equal to the given"? The implementation starts at `start_height` and moves forwards along the chain.